### PR TITLE
Fixes #37718 - BuildModal error message not copyable

### PR DESCRIPTION
**Pull request by Thorben-D**
[Redmine issue and reproducer](https://projects.theforeman.org/issues/37718)

This PR fixes a bug in the [BuildModal](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/BuildModal.js) component. 
The issue was that an error message was not copyable from the tree-view. This is fixed here by wrapping it in a `ClipboardCopy` component. 
Additionally, this PR removes the active element handling as it not needed and only interferes with the styling.

image

### CODE REVIEW

**Comment by sbernhard** [comment]
Thanks @Thorben-D for your suggestion. Actually, I would prefer to have one Copy to Clipboard for all found issues. Would this be doable? 
Thoughts on this @MariaAga?

**Comment by ekohl** [comment]
Wouldn't it be better if we linked to the exact template that failed? We may need to restructure the errors for that but that would be better to debug. Bonus points if you can find a line number

**Comment by sbernhard** [comment]
> Wouldn't it be better if we linked to the exact template that failed? We may need to restructure the errors for that but that would be better to debug. Bonus points if you can find a line number

Isn't this a different story? This is about a way to copy and paste error messages.

**Comment by MariaAga** [comment]
> Thanks @Thorben-D for your suggestion. Actually, I would prefer to have one Copy to Clipboard for all found issues. Would this be doable?
Thoughts on this @MariaAga?

Doable? yes with a button and `navigator.clipboard.writeText(...)`
I'm not sure if its more helpful to copy all issues and not one at a time, since in other contextes, when I have a list of error I google search them one by one and resolve them, one by one.

**Comment by sbernhard** [comment]
> > Thanks @Thorben-D for your suggestion. Actually, I would prefer to have one Copy to Clipboard for all found issues. Would this be doable?
> > Thoughts on this @MariaAga?
> 
> Doable? yes with a button and `navigator.clipboard.writeText(...)` I'm not sure if its more helpful to copy all issues and not one at a time, since in other contextes, when I have a list of error I google search them one by one and resolve them, one by one.

ok. I thoughts its maybe easier but I'm happy with both ways :)

**Review by MariaAga — CHANGES_REQUESTED** [review]
There are console error with this pr, I'll try to look if we can work around them somehow.
we cant put a button as the child of a tree item

**Comment by MariaAga** [comment]
Probably gonna need to change the tree to https://v4-archive.patternfly.org/v4/components/expandable-section to avoid errors (<button> cannot appear as a descendant of <button>.)

**Comment by ekohl** [comment]
> Isn't this a different story? This is about a way to copy and paste error messages.

Perhaps. As a user I'm surprised they're copyable now. Looks like an underdeveloped feature to me. And this is a new feature, so I'm going to give my thoughts on it.

Taking it a step back, I think I can derive that the user wants to resolve the issues. The question is: how do you make it actionable. To me it looks like you really want more structured errors. Taking apart the shown errors:

> Failure parsing Kickstart default iPXE: undefined method `interpolate_vars for nil:NilClass

This should link to `Kickstart default iPXE`. The real error is "undefined method \`interpolate_vars for nil:NilClass" and I'd argue that part should be visible. The way I would debug is to open the template (hence why you need the link) and look for `interpolate_vars`.

Technically I'd qualify this as a bug, but users can write their own templates so it's hard to figure out.

This would be much easier if it had a line number. I suspect this is lost in the Safemode renderer, but it's hard to tell.

> Failure parsing Kickstart default: ERF42-6371 [Foreman::Exception]: Neither disk nor partition table defined for host.

This looks like a "normal" exception because it has an error code. Those are calculated based on the class name and error message:
https://github.com/theforeman/foreman/blob/19b2129e0ab6fabb5585b3cd1c503e28edcaf685/lib/foreman/exception.rb#L8-L14

We have https://projects.theforeman.org/projects/foreman/wiki/ErrorCodes with some of these, though we haven't updated the list in a while. There's a rake task `exception:code` to generate a list of these codes (as the wiki page mentions). If you properly develop this feature, you can create a link to it. Our LinksController has a feature to link to the wiki.

Long term we can also look at including it in official docs so it can easily be branded as well.

> No smart proxies found

This is a bad error and not actionable at all. First of all, what kind of proxy does it look for? DHCP? DNS? TFTP? etc. Copy-pasting this error is IMHO useless.

Looks like it comes from here:
https://github.com/theforeman/foreman/blob/19b2129e0ab6fabb5585b3cd1c503e28edcaf685/app/services/host_build_status.rb#L44
That's not a very good error and I needed to read the actual code to figure out what it meant.

In conclusion: my feedback is that we need to do a better job of error handling.

**Comment by Thorben-D** [comment]
@MariaAga 
Thanks for noticing, this completely eluded me.
One option could be to turn the whole message into a copy button, as the leaves of the TreeView component, for some to me inexplicable reason, are apparently `<button>` elements.
I don't really like this approach though, because it feels more like a hack than clean UI, so I'll use a tree-table.
That is definitely overkill for now but, especially looking at  @ekohl's very valid suggestions, I believe that the added flexibility of the Table component will make expanding this errors section much easier in the future.

**Comment by Thorben-D** [comment]
@MariaAga I have updated the code and re-implemented the tree with a table. Errors are now gone.
It is quite a bit more code but, as stated above, I think it is worth it regarding the added flexibility.
Visually, not a lot has changed. I chose the "compact" style for the table, as I feel like having the lines looks nicer. Could be changed to "compact-borderless" though, if the old, line-less, style is preferred.

Could you have another look?

File: webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorsTreePrimaryRow.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        );
+        return isDetailsExpanded
+          ? otherDetailsExpandedNodeNames
+          : [...otherDetailsExpandedNodeNames, props.node.id];
+      }),
+    props: {
+      isExpanded,
+      isDetailsExpanded,
+      'aria-level': 1,
+      'aria-posinset': props.posinset,
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorsTreePrimaryRow.js:33]: Please add `posinset`,  `setsize` to `.eslintrc` skipWords

File: webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorPresenter.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { ClipboardCopy } from '@patternfly/react-core';
+import PropTypes from 'prop-types';
+
+export const ErrorPresenter = props => (
+  <ClipboardCopy
+    isReadOnly
+    hoverTip="Copy"
+    clickTip="Copied!"
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorPresenter.js]: Should be translated

File: webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/ErrorsTree.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      aria-label="Tree table"
+      variant="compact"
+    >
+      <Tbody>{renderPrimaryRows(props.data)}</Tbody>
+    </TableComposable>
+  );
+};
+
+ErrorsTree.propTypes = {
+  data: PropTypes.object,
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/ErrorsTree.js]: data is an array and not an object "checkPropTypes.js:20 Warning: Failed prop type: Invalid prop `data` of type `array` supplied to `ErrorsTree`, expected `object`."

File: webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorPresenter.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { ClipboardCopy } from '@patternfly/react-core';
+import PropTypes from 'prop-types';
+
+export const ErrorPresenter = props => (
+  <ClipboardCopy
+    isReadOnly
```
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorPresenter.js:8]: Can you please add `isBlock` I think it look nicer, it makes the grey background not have spaces between lines

**Review by MariaAga — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorPresenter.js (hunk 4 above)
Comment (Thorben-D) [reply webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorPresenter.js:8]: I agree, it looks much better. Added.

File: webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/ErrorsTree.js (hunk 3 above)
Comment (Thorben-D) [reply webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/ErrorsTree.js]: Fixed

File: webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorPresenter.js (hunk 2 above)
Comment (Thorben-D) [reply webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorPresenter.js]: Fixed

File: webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorsTreePrimaryRow.js (hunk 1 above)
Comment (Thorben-D) [reply webpack/assets/javascripts/react_app/components/HostDetails/ActionsBar/ErrorsTree/components/ErrorsTreePrimaryRow.js:33]: Added

**Comment by ekohl** [comment]
@Thorben-D is the screenshot in the opening message still up to date with the current code?

**Comment by Thorben-D** [comment]
@ekohl It is now ;)

**Comment by sbernhard** [comment]
@MariaAga Can you have a look at this again and hopefully merge this then?

**Review by MariaAga — APPROVED** [review]
LGTM, @ekohl Does the looks of it, look ok?

**Review by ekohl — COMMENTED** [review]
I think this looks OK (though I'm trusting @MariaAga on the proper JS review), but I think that if you rebase we shouldn't see the CI failures anymore. At least develop is reliably green now. Once green I have no objections to merging this, though I'd still encourage future improvements to make it even easier for users to figure out why it's failing.
